### PR TITLE
LibWasm+LibJS+LibWeb: Basic JavaScript API support for WebAssembly

### PR DESCRIPTION
**Pull request by alimpfard**
Lots of things are still a work in progress, but they're coherent enough to merge, I feel.
This also comes with a free Mandelbrot demo:
image_2021-05-22_23-40-45

Don't eat all those 38400 pixels all in one go!

### CODE REVIEW

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+            [&](const Wasm::FunctionAddress& address) {
+                auto function = create_native_function(address, export_.name(), global_object);
+                m_exports_object->define_property(export_.name(), function);
+            },
+            [&](const auto&) {
+                // FIXME: Implement other exports!
+            });
+    }
+
+    define_property("exports", m_exports_object);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:370]: This should be a getter instead (https://webassembly.github.io/spec/js-api/#dom-instance-exports), i.e.:

```cpp
define_native_property("exports", exports_getter, nullptr);
```

and then define a simple getter function that returns `m_exports_object`.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    virtual ~WebAssemblyObject() override = default;
+
+    struct CompiledWebAssemblyModule {
+        CompiledWebAssemblyModule(ByteBuffer&& source, Wasm::Module&& module)
+            : source(move(source))
+            , module(move(module))
+        {
+        }
+
+        ByteBuffer source; // This is kept around because the spec wants it at [[Source]]
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:30]: the spec doesn't care what we do with internal properties though, if this isn't used anywhere there's no need to keep it around.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 3]
```diff
@@ -89,16 +93,20 @@ JS_DEFINE_NATIVE_FUNCTION(WebAssemblyObject::instantiate)
...
-        vm.clear_exception();
-    }
     auto promise = JS::Promise::create(global_object);
-    if (!exception.is_empty()) {
-        promise->reject(exception);
+    auto take_exception_and_reject_if_needed = [&] {
+        if (vm.exception()) {
+            JS::Value exception = vm.exception()->value();
+            promise->reject(exception);
+            vm.clear_exception();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:107]: Let's make sure to clear the exception first, rejecting the promise can lead to JS function calls (not right now I think, but definitely once the `unhandledrejection` event is implemented), and that would fail miserably. :)

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 4]
```diff
@@ -117,25 +125,77 @@ JS_DEFINE_NATIVE_FUNCTION(WebAssemblyObject::instantiate)
...
-        auto promise = JS::Promise::create(global_object);
-        if (!exception.is_empty()) {
-            promise->reject(exception);
+        if (take_exception_and_reject_if_needed())
             return promise;
+
+        dbgln("Trying to resolve stuff because import object was specified");
+        for (const Wasm::Linker::Name& import_name : linker.unresolved_imports()) {
+            dbgln("Trying to resolve {}::{}", import_name.module, import_name.name);
+            auto value = import_object->get(import_name.module);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:145]: Needs an exception check, could throw via Proxy for example.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 5]
```diff
@@ -117,25 +125,77 @@ JS_DEFINE_NATIVE_FUNCTION(WebAssemblyObject::instantiate)
...
+            auto object = value.to_object(global_object);
+            if (vm.exception())
+                break;
+
+            auto import_ = object->get(import_name.name);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:150]: Needs an exception check

**Review by linusg — COMMENTED** [review]
Didn't nearly make it to the end yet, so here's a partial review instead. Exciting stuff! :partying_face:

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+{
+    ByteBuffer bytes;
+    if (is<JS::ArrayBuffer>(buffer)) {
+        auto array_buffer = static_cast<JS::ArrayBuffer*>(buffer);
+        bytes = array_buffer->buffer();
+    } else if (is<JS::TypedArrayBase>(buffer)) {
+        auto array = static_cast<JS::TypedArrayBase*>(buffer);
+        bytes = array->viewed_array_buffer()->buffer();
+    } else {
+        auto error = JS::TypeError::create(global_object, String::formatted("{} is not an ArrayBuffer", buffer->class_name()));
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:56]: That formatting capability is only available via `VM::throw_exception(GlobalObject&, ErrorType, Args...)`, the individual error classes do not have it.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:56]: oh, right. Unfortunate, but nevermind then.

[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:56]: I could add one, if you provide a name for it that doesn't sound terrible like `JS::Error::create_error<JS::TypeError>(...)`

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:56]: I think ideally it would just be an overload of `ErrorType::create()` if that's possible? I.e. either `TypeError::create(global_object, ErrorType,  Args...)` or `TypeError::create(global_object, const String&)`. Let's not get carried away though, we can deal with this separately.

[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:56]: Alright, we'll add it later then : )

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        builder.append("LinkError: Missing ");
+        builder.join(' ', link_result.error().missing_imports);
+        auto error = JS::TypeError::create(global_object, builder.build());
+        promise->reject(error);
+        return promise;
+    }
+
+    auto instance_result = s_abstract_machine.instantiate(*module, link_result.release_value());
+    if (instance_result.is_error()) {
+        auto error = JS::TypeError::create(global_object, instance_result.error().error);
```
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:227]: so, what's the answer? :D

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    virtual ~WebAssemblyModuleObject() override = default;
+
+    size_t index() const { return m_index; }
+    const Wasm::Module& module() const { return WebAssemblyObject::s_compiled_modules.at(m_index).module; }
+
+private:
+    size_t m_index { 0 };
+};
+
+class WebAssemblyModuleInstanceObject final : public JS::Object {
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:61]: Maybe this should be called `WebAssemblyInstanceObject` (`WebAssembly.Instance`), to mirror `WebAssemblyModuleObject` -> `WebAssembly.Module`?

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+public:
+    explicit WebAssemblyMemoryObject(JS::GlobalObject&, Wasm::MemoryAddress);
+    virtual void initialize(JS::GlobalObject&) override;
+    virtual ~WebAssemblyMemoryObject() override = default;
+
+    auto address() const { return m_address; }
+
+private:
+    // JS_DECLARE_NATIVE_FUNCTION(grow);
+    JS_DECLARE_NATIVE_GETTER(buffer);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:93]: Totally forgot this previously, this needs to have a distinct prototype (not `global_object.object_prototype()`) and not be on the instance directly (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Instance/exports) - sorry!

This should be relatively easy, but if you need any help just let me know :)

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    m_exports_object->set_integrity_level(IntegrityLevel::Frozen);
+
+    define_native_property("exports", exports_getter, nullptr);
+}
+
+JS_DEFINE_NATIVE_GETTER(WebAssemblyModuleInstanceObject::exports_getter)
+{
+    auto this_value = vm.this_value(global_object);
+    if (vm.exception())
+        return {};
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:387]: `this_value()` can't throw but `to_object()` below can (in which case we would override the exception by throwing again but shouldn't), so just move this one line down :)

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+}
+
+JS_DEFINE_NATIVE_FUNCTION(WebAssemblyObject::instantiate)
+{
+    // FIXME: This shouldn't block!
+    auto buffer = vm.argument(0).to_object(global_object);
+    auto promise = JS::Promise::create(global_object);
+    auto take_exception_and_reject_if_needed = [&] {
+        if (vm.exception()) {
+            JS::Value rejection_value = vm.exception()->value();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp:105]:
```suggestion
            auto rejection_value = vm.exception()->value();
```

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 12]
```diff
@@ -27,7 +32,13 @@ ArrayBuffer::ArrayBuffer(size_t byte_size, Object& prototype)
 
 ArrayBuffer::ArrayBuffer(ByteBuffer& buffer, Object& prototype)
     : Object(prototype)
-    , m_buffer(buffer)
+    , m_buffer(ByteBuffer { buffer })
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:35]: As discussed on Discord earlier - this can probably just be

```suggestion
    , m_buffer(buffer)
```

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 13]
```diff
@@ -27,7 +32,13 @@ ArrayBuffer::ArrayBuffer(size_t byte_size, Object& prototype)
...
+{
+}
+
+ArrayBuffer::ArrayBuffer(ByteBuffer* buffer, Object& prototype)
+    : Object(prototype)
+    , m_buffer(static_cast<ByteBuffer*>(buffer))
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:41]: `static_cast`'ing a `ByteBuffer*` to `ByteBuffer*` might not be needed either?

```suggestion
    , m_buffer(buffer)
```

File: Base/res/html/misc/mandelbrot-wasm.html [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+                        0x3b, 0x01, 0x00, 0x20, 0x0e, 0x41, 0x01, 0x6a, 0x21, 0x0e, 0x0c, 0x01,
+                        0x0b, 0x0b, 0x20, 0x0d, 0x41, 0x01, 0x6a, 0x21, 0x0d, 0x0c, 0x01, 0x0b,
+                        0x0b, 0x0b, 0x07, 0x00, 0x20, 0x00, 0x2f, 0x01, 0x00, 0x0b, 0x00, 0x17,
+                        0x10, 0x73, 0x6f, 0x75, 0x72, 0x63, 0x65, 0x4d, 0x61, 0x70, 0x70, 0x69,
+                        0x6e, 0x67, 0x55, 0x52, 0x4c, 0x05, 0x66, 0x61, 0x6c, 0x73, 0x65,
+                    ])
+                );
+            })
+                .then(bytes => {
+                    return WebAssembly.compile((bytess = bytes));
```
Comment (linusg) [inline Base/res/html/misc/mandelbrot-wasm.html:428]: wait, what's the purpose of that?

Comment (alimpfard) [reply Base/res/html/misc/mandelbrot-wasm.html:428]: Giving me a way to take a look at it lol
There's a "exports = module.exports" below as well

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h (hunk 9 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:93]: Oh I see (I assume you meant to comment on `JS_DECLARE_NATIVE_GETTER(exports_getter)`)
so I should make a `WebAssemblyInstancePrototype` and put these on that instead I guess?

Also this comment is annoying, I'll get rid of it or implement it

File: Base/res/html/misc/mandelbrot-wasm.html (hunk 14 above)
Comment (alimpfard) [reply Base/res/html/misc/mandelbrot-wasm.html]: I fixed these up to no longer assign to random global variables.

File: Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h (hunk 9 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:108]: > I assume you meant to comment on JS_DECLARE_NATIVE_GETTER(exports_getter)

oops, yes!

> so I should make a WebAssemblyInstancePrototype and put these on that instead I guess?

Precisely 👍

Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:108]: Hmmm
How do I make all of them share one prototype object?

Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:108]: Basically like this:

https://github.com/SerenityOS/serenity/blob/bdd7741ae12955bd31225870fa85e96a6e79e9dd/Userland/Libraries/LibWeb/Bindings/WindowObjectHelper.h#L227-L233

https://github.com/SerenityOS/serenity/blob/1682f0b760c736e7bbb62bea86248bd83579bec7/Userland/Libraries/LibWeb/Bindings/WindowObject.h#L35-L56

Which allows you to get the unique prototype like this:

https://github.com/SerenityOS/serenity/blob/bdd7741ae12955bd31225870fa85e96a6e79e9dd/Userland/Libraries/LibWeb/CodeGenerators/WrapperGenerator.cpp#L904

TL;DR: Create a prototype once, add it to the global object's `m_prototypes` to save it from GC, set it on `WebAssemblyInstanceObject`. You don't have to use these specific functions, they're mostly for the generated wrapper code.

Comment (alimpfard) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:108]: I can't just insert random stuff into the global object though, this is inserted from LibWeb, and I don't have access to the WindowObject directly (I can probably cast the global object to that), but that's just ugly...

Comment (linusg) [reply Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h:108]: `static_cast<WindowObject&>(global_object)` is exactly how we do this :)
